crates/trie/trie/src/trie_cursor/mod.rs

```diff
@@ -52,3 +53,29 @@ pub trait TrieCursor: Send + Sync {
     /// Get the current entry.
     fn current(&mut self) -> Result<Option<Nibbles>, DatabaseError>;
 }
+
+/// A trait for iterating over a range of block numbers.
+#[auto_impl::auto_impl(&mut, Box)]
+pub trait TrieRangeWalker<V>: Send + Sync {
+    /// Creates an iterator that walks over a range of block numbers.
+    fn walk_range(
+        &mut self,
+        range: impl RangeBounds<BlockNumber>,
+    ) -> Result<impl Iterator<Item = Result<V, DatabaseError>>, DatabaseError>;
+}
+
+/// Creates readable trie cursors that iterate over a range of block numbers.
+#[auto_impl::auto_impl(&mut, Box)]
+pub trait TrieRangeWalkerFactory: Send + Sync {
```

Comment:
think this name feels a bit confusing: we don't walk a trie range, but rather a changesets range
```suggestion
pub trait ChangeSetsWalkerFactory: Send + Sync {
```